File: bfbs.pl
#!/usr/bin/perl
#
# B F B S . P L
#
# bfbs.pl last edited on Sat Feb 28 17:17:13 2026 as version 0v9
#
# This script reads one or more CSV files, containing one or more columns of
# numbers and calculates basic statistics for each column (including sum,
# average, variance and standard deviation) and outputs the results.
#

# ChatGPT prompts that produced this bfbs.pl script; -
# 1. Please write a perl script that reads one or more 
#    comma separated value data files containing one or more
#    columns of numbers and calculates the sum, mean,
#    variance and standard deviation for each column
#    using the arbitrary-precision floating point
#    arithmetic package Math::BigFloat and writes
#    out the results for each column in each file.
#    Ensure that the script can handle leading spaces
#    on numbers, blank rows in files and comment lines
#    beginning with a hash. Allow the precision to be
#    controlled by the user from the command line.
# 2. Please make the calculated variance the sample form
#    of variance i.e. = sum((x - mean)^2) / (n - 1)
#    unless the user sets an option in the command line.
# 3. Please enhance the script to calculate and output
#    min, median, max and range.
# 4. Please enhance the script to handle headers in the
#    top of row of the CSV columns when the user flags
#    this in the command line.
# 5. Please add a command line option to make the output
#    use scientific format e.g. 1.0e1
# 6. ~ Fix errors
# 7. Please add announcement info at the start of output
#    that has the program name and version number,
#    then the version of perl that is running the code and
#    then the version numbers of the packages used if they
#    exist and finally output the precision and output format
#    being used.

#
# 0v9 Added population variance and population standard deviation calculations and output
# 0v8 Add --quiet option to suppress elapsed time output
# 0v7 Add code execution elapsed calculation
# 0v6 Cosmetic changes to output labels

# Deviations from other bfbs implementations; -
# 1. This Perl version of bfbs aborts with an error if the
# input file cannot be opened, whereas the Ruby and Python
# versions print a warning and skip to the next file, if there
# is a next file.

# Perl Math::BigFloat's precision method sets the
# number of significant digits, not decimal places.
# This 0v8 version of bfbs.pl uses the "accuracy"
# method to set the number of digits each result
# should have, which is more intuitive for users
# who want to specify precision in terms of digits.
# The "precision" method in Math::BigFloat sets the
# number of significant digits, which can lead to
# confusion when dealing with numbers of varying
# magnitudes. By using "accuracy", we ensure that
# all results are consistently formatted to the
# specified number of digits, regardless of their size.
# (see https://perldoc.perl.org/Math::BigFloat)

# MARK: 
# === Modules ===
use strict;
use warnings;
use Getopt::Long;
use Math::BigFloat;     # arbitrary-precision floating point arithmetic
use Time::HiRes qw(time);   # elapsed time measurement

# === Start Timing ===
my $start_time = time();

# === Program Info ===
my $PROGRAM_NAME    = "bfbs.pl";
my $PROGRAM_VERSION = "0v9 (2026-04-14)";

# MARK:
# === Settings ===
my $precision       = 40;
my $use_header      = 0;
my $use_scientific  = 0;
my $quiet           = 0;
my $help            = 0;
my $helpMsg         = << "HELP_MSG";
[--header] [--help] [--precision=N] [--scientific] [--quiet] file1.csv [file2.csv ...]
 where:
  --header       : Treat first CSV row as column headers
  --help         : Show this help message and exit
  --precision=N  : Set precision to N digits (default: 40)
  --scientific   : Output numbers in scientific notation (e.g. 1.0e1)
  --quiet        : Suppress version and elapsed time info output
  file1.csv ...  : One or more CSV files to process
HELP_MSG

GetOptions(
    "precision=i"   => \$precision,
    "header"        => \$use_header,
    "help"          => \$help,
    "scientific"    => \$use_scientific,
    "quiet"         => \$quiet,
) or die "Usage: $0 $helpMsg\n";

# Show help and exit if requested
if ($help) { die "Usage: $0 $helpMsg\n" };

# Set global Math::BigFloat precision
Math::BigFloat->accuracy($precision);  # "accuracy()" sets the number of digits each result should have

# Require at least one file
@ARGV or die "Error: No input files specified.\nUsage: $0 $helpMsg\n";

# === Announce Info ===
unless ($quiet) {
    print "$PROGRAM_NAME version $PROGRAM_VERSION\n";
    print "Perl version: $^V\n";
    print "Getopt::Long Version: ", ($Getopt::Long::VERSION // 'unknown'), "\n";
    print "Math::BigFloat Version: ", ($Math::BigFloat::VERSION // 'unknown'), "\n";
}
print "Precision: $precision digits, Output Format: ", ($use_scientific ? "Scientific" : "Decimal"), "\n";

# MARK:
# === File Processing ===
foreach my $file (@ARGV) {
    open my $fh, '<', $file or die "Error: Cannot open \"$file\": $!";

    print "\nFile: $file\n";

    my @columns;       # Array of arrayrefs for column data
    my @headers;       # Column headers
    my $header_found = 0;

    while (my $line = <$fh>) {
        chomp $line;
        $line =~ s/\r//g;

        next if $line =~ /^\s*$/;    # Blank lines
        next if $line =~ /^\s*#/;    # Comments

        my @fields = split /\s*,\s*/, $line;

        if ($use_header && !$header_found) {
            @headers = @fields;
            $header_found = 1;
            next;
        }

        for my $i (0 .. $#fields) {
            my $val = $fields[$i];
            $val =~ s/^\s+//;

            $columns[$i] ||= [];
            push @{$columns[$i]}, Math::BigFloat->new($val);
        }
    }

    close $fh;

    for my $i (0 .. $#columns) {
        my $data = $columns[$i];
        next unless @$data;

        my $label = $use_header ? ($headers[$i] // ($i + 1)) : ($i + 1);

        my $n = scalar @$data;
        my $sum = Math::BigFloat->new(0);
        $sum->badd($_) for @$data;

        my $mean = $sum->copy()->bdiv($n);

        # Variance
        my $var_sum = Math::BigFloat->new(0);
        for my $x (@$data) {
            my $diff = $x->copy()->bsub($mean);
            $var_sum->badd($diff->bpow(2));
        }

        my $ndivisor;
        my $n1divisor;
        if ($n <= 1) {
            warn "Not enough data in $label to compute sample variance (n = $n)\n";
            next;
        }
        $ndivisor = Math::BigFloat->new($n);
        $n1divisor = Math::BigFloat->new($n - 1);

        my $pvariance = $var_sum->copy()->bdiv($ndivisor);
        my $pstddev   = $pvariance->copy()->bsqrt();
        my $svariance = $var_sum->copy()->bdiv($n1divisor);
        my $sstddev   = $svariance->copy()->bsqrt();

        # Min, Median, Max, Range
        my @sorted = sort { $a->bcmp($b) } @$data;
        my $min    = $sorted[0]->copy();
        my $max    = $sorted[-1]->copy();
        my $range  = $max->copy()->bsub($min);
        my $median;

        if ($n % 2 == 1) {
            $median = $sorted[int($n / 2)]->copy();
        } else {
            my $mid1 = $sorted[$n/2 - 1]->copy();
            my $mid2 = $sorted[$n/2]->copy();
            $median = $mid1->badd($mid2)->bdiv(2);
        }

        print "\nColumn: $label\n";
        print "  Count        : $n\n";
        print "  Minimum      : ", ($use_scientific ? $min->bnstr() : $min->bstr()), "\n";
        print "  Mean         : ", ($use_scientific ? $mean->bnstr() : $mean->bstr()), "\n";
        print "  Median       : ", ($use_scientific ? $median->bnstr() : $median->bstr()), "\n";
        print "  Maximum      : ", ($use_scientific ? $max->bnstr() : $max->bstr()), "\n";
        print "  Range        : ", ($use_scientific ? $range->bnstr() : $range->bstr()), "\n";
        print "  Sum          : ", ($use_scientific ? $sum->bnstr() : $sum->bstr()), "\n";
        print "  Variance (s²): ", ($use_scientific ? $svariance->bnstr() : $svariance->bstr()), "\n";
        print "  Std. Dev. (s): ", ($use_scientific ? $sstddev->bnstr() : $sstddev->bstr()), "\n";
        print "  Variance (σ²): ", ($use_scientific ? $pvariance->bnstr() : $pvariance->bstr()), "\n";
        print "  Std. Dev. (σ): ", ($use_scientific ? $pstddev->bnstr() : $pstddev->bstr()), "\n";
    }
}

# === Elapsed Time Calculation ===
unless ($quiet) {
    my $end_time = time();
    my $elapsed = $end_time - $start_time;
    printf("\nbfbs.pl execution elapsed time: %.6f [sec]\n", $elapsed);
}

File: bfbs.ts
#!/usr/bin/env -S deno run --allow-read

//
// B F B S . T S
//
// bfbs.ts last edited on Sat Mar 28 23:50:51 2026 as version 0.0.3
//

// A Deno script to compute statistics from CSV files with arbitrary precision.

// bfbs.ts originated from free ChatGPT code and is licensed
// under the MIT License. The original prompt was; -
// Please write a javascript script to be executed by deno
// (i.e. command line) that reads one or more comma separated
// value data files containing one or more columns of numbers
// and calculates the min, median, mean, max, range, sum,
// variance and standard deviation for each column using an
// arbitrary-precision floating point arithmetic package and
// writes out the results for each column in each file. Ensure
// that the script can handle leading spaces on numbers, blank
// rows in files and comment lines beginning with a hash. Allow
// the precision to be controlled by the user from the command
// line. Please make the calculated variance the sample form of
// variance i.e. = sum((x - mean)^2) / (n - 1) unless the user
// sets an option in the command line. Please enhance the script
// to handle headers in the top of row of the CSV columns when
// the user flags this in the command line.

// Usage/Help message is; -
// Usage: deno run --allow-read bfbs.ts [options] <file1.csv> [file2.csv...]
// Options:
//   --precision N or -P N   Set decimal precision to N (default is 40)
//   --header or -H   First row contains column headers
//   --help or -h     Show this help message
//   --quiet or -q    Suppress some output information

// 0.0.3 - Check for unknown options. Calc both sample and population
//         variance / standard deviation. Removed --population option.
// 0.0.2 - Added --quiet option to suppress some output information.
// 0.0.1 - Initial version.

import Decimal from "https://esm.sh/decimal.js@10.4.3";

// MARK: Command line parsing

// --------------------
// Command line parsing
// --------------------

interface Options {
  precision: number;
  header: boolean;
  population: boolean;
  quiet: boolean;
  help: boolean;
  files: string[];
}

function parseArgs(args: string[]): Options {
  const options: Options = {
    precision: 40,
    header: false,
    population: false,
    quiet: false,
    help: false,
    files: [],
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];

    if ((arg === "--precision" || arg === "-P") && args[i + 1]) {
      options.precision = parseInt(args[++i], 10);
    } else if (arg === "--header" || arg === "-H") {
      options.header = true;
    } else if (arg === "--quiet" || arg === "-q") {
      options.quiet = true;
    } else if (arg === "--help" || arg === "-h") {
      options.help = true;
    } else if (arg.slice(0, 2) === "--") {
      console.error(`Warning: Ignoring unknown long option: ${arg}`);
    } else if (arg.slice(0, 1) === "-") {
      console.error(`Warning: Ignoring unknown short option: ${arg}`);
    }  else {
      options.files.push(arg);
    }
  }

  if (options.files.length === 0 || options.help ) {
    console.error("Usage: deno run --allow-read bfbs.ts [options] <file1.csv> [file2.csv...]");
    console.error("Options:");
    console.error("  --precision N or -P N   Set decimal precision to N (default is 40)");
    console.error("  --help or -h     Show this help message");
    console.error("  --header or -H   First row contains column headers");
    console.error("  --quiet or -q    Suppress some output information");
    Deno.exit(1);
  }

  return options;
}

const options = parseArgs(Deno.args);

// Configure decimal precision
Decimal.set({ precision: options.precision });

// MARK: Utility functions

// -----------------
// Utility functions
// -----------------

function median(values: Decimal[]): Decimal {
  const sorted = [...values].sort((a, b) => a.comparedTo(b));
  const n = sorted.length;

  if (n === 0) return new Decimal(0);

  if (n % 2 === 0) {
    return sorted[n / 2 - 1].plus(sorted[n / 2]).div(2);
  } else {
    return sorted[Math.floor(n / 2)];
  }
}

function computeStats(values: Decimal[]) {
  const n = values.length;

  if (n === 0) return null;

  const sum = values.reduce((acc, v) => acc.plus(v), new Decimal(0));
  const mean = sum.div(n);

  let variance = new Decimal(0);
  let pvariance = new Decimal(0);

  if (n > 0) {
    const squaredDiffs = values.map(v => v.minus(mean).pow(2));
    const numerator = squaredDiffs.reduce((acc, v) => acc.plus(v), new Decimal(0));
    if (n > 1) {
      variance = numerator.div(n - 1);
    }
    pvariance = numerator.div(n);
  }

  const stddev = variance.sqrt();
  const pstddev = pvariance.sqrt();

  const min = values.reduce((a, b) => Decimal.min(a, b));
  const max = values.reduce((a, b) => Decimal.max(a, b));
  const range = max.minus(min);

  return {
    count: n,
    sum,
    mean,
    median: median(values),
    min,
    max,
    range,
    variance,
    stddev,
    pvariance,
    pstddev
  };
}

// MARK: Process a File

// -------------------
// CSV File Processing
// -------------------

async function processFile(filename: string) {
  const text = await Deno.readTextFile(filename);
  const lines = text.split(/\r?\n/);

  let headers: string[] = [];
  const columns: Decimal[][] = [];

  let firstDataRow = true;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (line === "") continue;
    if (line.startsWith("#")) continue;

    const parts = rawLine.split(",");

    if (firstDataRow && options.header) {
      headers = parts.map(h => h.trim());
      firstDataRow = false;
      continue;
    }

    firstDataRow = false;

    parts.forEach((value, i) => {
      const trimmed = value.trim();
      if (trimmed === "") return;

      try {
        const num = new Decimal(trimmed);
        if (!columns[i]) columns[i] = [];
        columns[i].push(num);
      } catch {
        // Ignore non-numeric values
      }
    });
  }

  console.log(`\nProcessing file: ${filename}`);

  columns.forEach((col, i) => {
    const stats = computeStats(col);
    if (!stats) return;

    const name = options.header && headers[i]
      ? headers[i]
      : `Column: ${i + 1}`;

    console.log(`\n${name}`);
    console.log(`Count.       : ${stats.count}`);
    console.log(`Min.         : ${stats.min.toString()}`);
    console.log(`Mean.        : ${stats.mean.toString()}`);
    console.log(`Median.      : ${stats.median.toString()}`);
    console.log(`Max.         : ${stats.max.toString()}`);
    console.log(`Range.       : ${stats.range.toString()}`);
    console.log(`Sum.         : ${stats.sum.toString()}`);
    console.log(`Variance (s²): ${stats.variance.toString()}`);
    console.log(`Std. Dev. (s): ${stats.stddev.toString()}`);
    console.log(`Variance (σ²): ${stats.pvariance.toString()}`);
    console.log(`Std. Dev. (σ): ${stats.pstddev.toString()}`);
  });
}

// MARK: Main processing loop

// ------------------------
// Run main processing loop
// ------------------------

const start = performance.now();
if (!options.quiet) {
  console.log("bfbs.ts version 0.0.3");
  console.log(`Processing files with ${options.precision} digits of precision `);
}

for (const file of options.files) {
  try {
    await processFile(file);
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) {
      console.error(`\nWarning: Unable to find File named "${file}", as it does not exist.`);
    } else if (error instanceof Deno.errors.PermissionDenied) {
      console.error(`\nWarning: Permission denied for file named "${file}" on read attempt.`);
    } else if (error instanceof Deno.errors.IsADirectory) {
      console.error(`\nWarning: Skipping file named "${file}" as it is a directory.`);
    } else {
      throw error;
    }
  }
}

if (!options.quiet) {
  const end = performance.now();
  console.log(`\nbfbs.ts processing took ${(end - start).toFixed(2)} [mS].`);
}
